server/core/src/rodsServer.cpp

```diff
@@ -1385,6 +1385,31 @@ int initServerMain(
 
     const auto start_delay_server = [&](std::any&)
     {
+
+
+/*  NEW 4.3.0 LEADER/SUCCESSOR ALGORITHM
+
+   - if self is leader
+     - if successor is defined
+        - if necessary, gracefully complete current jobs and exit
+     - else
+        - if necessary, start delay server
+        - grab and process delay rules, same as today
+   - else if self is successor
+     - run health check on the leader stats (utility function that calculates based on db stats)
+     - if leader is dead
+         - promote self to leader in db
+         - exit
+     - else
+         - check on the leader directly (ping?  grid-status?)
+         - populate stats in the db (JSON?)
```

Comment:
should live in `r_grid_configuration` as `'delay_server', 'statistics', JSON`

might need another two admin APIs...  `append_delay_server_statistics` and `clear_delay_server_statistics`

value is a json array of...
- timestamp
- leader hostname
- pinger hostname
- ping time in milliseconds(?) or FAIL (or this is two different fields?)

Reply:
a couple candidates...

```
{
	[
		{
			'leader': 'awesome.example.org',
			'successor': 'thenextone.example.org',
			'timestamp': 1643580156,
			'milliseconds': 324
		},
		{
			'leader': 'awesome.example.org',
			'successor': 'thenextone.example.org',
			'timestamp': 1643580158,
			'milliseconds': 324
		},
		{
			'leader': 'awesome.example.org',
			'successor': 'thenextone.example.org',
			'timestamp': 1643580160,
			'milliseconds': 324
		},
		{
			'leader': 'awesome.example.org',
			'successor': 'thenextone.example.org',
			'timestamp': 1643580162,
			'milliseconds': 324
		},
		{
			'leader': 'awesome.example.org',
			'successor': 'thenextone.example.org',
			'timestamp': 1643580164,
			'milliseconds': 324
		}
	]
}
```

or

```
{
	'awesome.example.org': {
		'successor': 'thenextone.example.org',
		'events': [
			{
		        'timestamp': 1643580156,
		        'milliseconds': 324
			},
			{
		        'timestamp': 1643580158,
		        'milliseconds': 324
			},
			{
		        'timestamp': 1643580160,
		        'milliseconds': 324
			},
			{
		        'timestamp': 1643580162,
		        'milliseconds': 324
			},
			{
		        'timestamp': 1643580164,
		        'milliseconds': 324
			}
		]
	}
}
```

Reply:
or, for now, this can just be a singleton JSON string in the server.    no need to keep this in the database - only the same server (current successor) will access its own stats.

willing to revisit later, and get it into the database for greater visibility.